src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,140 @@
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "pandaproxy/kafka_client_cache.h"
+
+#include "pandaproxy/logger.h"
+
+#include <seastar/core/loop.hh>
+
+namespace pandaproxy {
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg,
+  config::binding<size_t> max_size,
+  config::binding<std::chrono::milliseconds> keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{keep_alive} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(std::move(user.name));
+        cfg.scram_password.set_value(std::move(user.pass));
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+ss::future<client_ptr> kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    ss::sstring k{user.name};
+    auto it_hash = inner_hash.find(k);
+
+    client_ptr client;
+
+    // When no client is found ...
+    if (it_hash == inner_hash.end()) {
+        if (_cache.size() >= _cache_max_size()) {
+            if (!_cache.empty()) {
+                auto item = inner_list.back();
+                co_await item.client->stop();
+                vlog(plog.debug, "Cache size reached, evicting {}", item.key);
+                inner_list.pop_back();
+            }
+        }
+
+        vlog(plog.debug, "Make client for user {}", k);
+
+        client = make_client(user, authn_method);
+        inner_list.emplace_front(k, client);
+    } else {
+        // Grab the node located at that position
+        auto node = inner_hash.extract(it_hash);
+
+        // Reuse the client if the passwords match
+        // Otherwise make a new client
+        if (node.value().client->config().scram_password.value() == user.pass) {
+            vlog(plog.debug, "Reuse client for user {}", k);
+        } else {
+            vlog(plog.debug, "Password changed. Remake client for user {}", k);
+            client_ptr old = node.value().client;
+            bool stop_old_client = false;
+            client_ptr next;
+            try {
+                next = make_client(user, authn_method);
+                // Try connect as a safety check
+                co_await next->connect();
+            } catch (const std::exception& ex) {
+                node.value().client = next;
+                stop_old_client = true;
```

Comment:
If I don't call stop then the client will remain connected and will be dangling after the new one is made.
I think instead of calling stop I can assign the old client to a temp user and let GC take it from there.

Reply:
If the new client fails, why would you replace what's in the cache and stop the old one?

I expect the behaviour to be for the new one to replace the old one, if and only if the new one connects successfully.

Reply:
Oh I did the wrong thing! ....
Fixing now.